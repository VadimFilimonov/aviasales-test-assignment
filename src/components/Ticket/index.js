/* eslint-disable react/prop-types */
import React from 'react';
import uuid from 'react-uuid';
import styled from 'styled-components';

const Wrapper = styled.article`
  display: flex;
  flex-wrap: wrap;
  align-items: center;
  justify-content: space-between;
  padding: 20px;
  background: #fff;
  border-radius: 5px;
  box-shadow: 0px 2px 8px rgba(0, 0, 0, .1);
`;

const Price = styled.div`
  color: #2196f3;
  font-weight: 600;
  font-size: 24px;
`;

const InfoList = styled.dl`
  display: grid;
  grid-template-columns: repeat(3, 1fr);
  width: 100%;
  font-weight: 600;
  column-gap: 20px;
`;

const InfoTitle = styled.dt`
  color: #a0b0b9;
  font-size: 12px;
  letter-spacing: .5px;
  text-transform: uppercase;
`;

const InfoDescription = styled.dd`
  margin: 0;
  color: #4a4a4a;
  font-size: 14px;
`;

const Ticket = ({ ticket }) => (
  <Wrapper>
    <Price>
      {ticket.price}
      Р
    </Price>
    <img src={`http://pics.avs.io/99/36/${ticket.carrier}.png`} alt="" />
    {ticket.segments.map((segment) => (
      <InfoList key={uuid()}>
        <div>
          <InfoTitle>{[segment.origin, segment.destination].join(' - ')}</InfoTitle>
          {/* TODO: нормализовать. Из timestamp в человеко-читабельную дату */}
          <InfoDescription>{segment.date}</InfoDescription>
        </div>
        <div>
          <InfoTitle>В пути</InfoTitle>
          {/* TODO: нормализовать. Например: 90 -> 1ч 30м */}
          <InfoDescription>{segment.duration}</InfoDescription>
        </div>
        <div>
          {/* TODO: Не выводить если нет пересадок; Проверить нужна ли плюрализация */}
          <InfoTitle>
            {segment.stops.length}
            пересадки
          </InfoTitle>
          <InfoDescription>{segment.stops.join(', ')}</InfoDescription>
        </div>
      </InfoList>
    ))}
  </Wrapper>
);

export default Ticket;
