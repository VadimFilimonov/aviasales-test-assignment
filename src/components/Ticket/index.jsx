/* eslint-disable react/prop-types */
import React from 'react';
import uuid from 'react-uuid';
import styled from 'styled-components';
import { convertToTime, convertToDuration } from '../../utilities/date';

const formatPrice = (price) => (
  new Intl.NumberFormat('ru-RU', { style: 'currency', currency: 'RUB' }).format(price).replace(',00', '')
);

const Wrapper = styled.article`
  display: flex;
  flex-wrap: wrap;
  align-items: center;
  justify-content: space-between;
  padding: 20px;
  background: ${(props) => props.theme.color.white};
  border-radius: 5px;
  box-shadow: 0px 2px 8px rgba(0, 0, 0, .1);
`;

const Price = styled.div`
  color: ${(props) => props.theme.color.blue};
  font-weight: 600;
  font-size: 24px;
`;

const Logo = styled.img`
  max-width: 110px;
  margin-right: 30px;

  @media screen and (max-width: 700px) {
    margin-right: 0;
  }
`;

const InfoList = styled.dl`
  display: grid;
  grid-template-columns: repeat(3, 1fr);
  width: 100%;
  font-weight: 600;
  column-gap: 20px;

  @media screen and (max-width: 700px) {
    column-gap: 5px;
  }
`;

const InfoTitle = styled.dt`
  color: ${(props) => props.theme.color.gullGray};
  font-size: 12px;
  letter-spacing: .5px;
  text-transform: uppercase;

  @media screen and (max-width: 700px) {
    font-size: 10px;
  }
`;

const InfoDescription = styled.dd`
  margin: 0;
  color: ${(props) => props.theme.color.black};
  font-size: 14px;

  @media screen and (max-width: 700px) {
    font-size: 12px;
  }
`;

const Ticket = ({ ticket }) => (
  <Wrapper>
    <Price>{formatPrice(ticket.price)}</Price>
    <Logo src={`https://pics.avs.io/99/36/${ticket.carrier}.png`} alt="" />
    {ticket.segments.map((segment) => (
      <InfoList key={uuid()}>
        <div>
          <InfoTitle>{[segment.origin, segment.destination].join(' - ')}</InfoTitle>
          <InfoDescription>{convertToTime(segment.date)}</InfoDescription>
        </div>
        <div>
          <InfoTitle>В пути</InfoTitle>
          <InfoDescription>{convertToDuration(segment.duration)}</InfoDescription>
        </div>
        {/* TODO: при отсутствии пересадок в description написать "Прямой" */}
        {segment.stops.length > 0
          && (
          <div>
            <InfoTitle>
              {/* TODO: add plural */}
              {[segment.stops.length, 'пересадки'].join(' ')}
            </InfoTitle>
            <InfoDescription>{segment.stops.join(', ')}</InfoDescription>
          </div>
          )}
      </InfoList>
    ))}
  </Wrapper>
);

export default Ticket;
