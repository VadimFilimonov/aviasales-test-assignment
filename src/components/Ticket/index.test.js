import React from 'react';
import { render } from '@testing-library/react';
import Ticket from './index';

test('Ticket 1', () => {
  const ticket = {
    carrier: 'EY',
    price: 44327,
    segments: [
      {
        date: '2020-12-12T11:24:00.000Z',
        destination: 'HKT',
        duration: 1171,
        origin: 'MOW',
        stops: ['SHA', 'BKK', 'DXB'],
      },
      {
        date: '2021-01-01T06:06:00.000Z',
        destination: 'MOW',
        duration: 702,
        origin: 'HKT',
        stops: [],
      },
    ],
  };
  // eslint-disable-next-line react/jsx-props-no-spreading
  const { asFragment } = render(<Ticket ticket={ticket} />);
  expect(asFragment()).toMatchSnapshot();
});
