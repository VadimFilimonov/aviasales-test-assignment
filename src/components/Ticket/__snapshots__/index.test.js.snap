// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Ticket 1 1`] = `
<DocumentFragment>
  <article
    class="sc-bdfBwQ iqAcWp"
  >
    <div
      class="sc-gsTCUz hdYghD"
    >
      44 327 ₽
    </div>
    <img
      alt=""
      class="sc-dlfnbm jjvIZi"
      src="http://pics.avs.io/99/36/EY.png"
    />
    <dl
      class="sc-hKgILt djIyHG"
    >
      <div>
        <dt
          class="sc-eCssSg djFfwt"
        >
          MOW - HKT
        </dt>
        <dd
          class="sc-jSgupP cMTqah"
        >
          14:24
        </dd>
      </div>
      <div>
        <dt
          class="sc-eCssSg djFfwt"
        >
          В пути
        </dt>
        <dd
          class="sc-jSgupP cMTqah"
        >
          19ч 31м
        </dd>
      </div>
      <div>
        <dt
          class="sc-eCssSg djFfwt"
        >
          3 пересадки
        </dt>
        <dd
          class="sc-jSgupP cMTqah"
        >
          SHA, BKK, DXB
        </dd>
      </div>
    </dl>
    <dl
      class="sc-hKgILt djIyHG"
    >
      <div>
        <dt
          class="sc-eCssSg djFfwt"
        >
          HKT - MOW
        </dt>
        <dd
          class="sc-jSgupP cMTqah"
        >
          09:06
        </dd>
      </div>
      <div>
        <dt
          class="sc-eCssSg djFfwt"
        >
          В пути
        </dt>
        <dd
          class="sc-jSgupP cMTqah"
        >
          11ч 42м
        </dd>
      </div>
    </dl>
  </article>
</DocumentFragment>
`;
